src/sideBar.ts

```diff
@@ -9,193 +18,238 @@ nls.config({ messageFormat: nls.MessageFormat.bundle, bundleFormat: nls.BundleFo
     }
     dispose(): void {
         this.sideBar.dispose;
     }
 }
 
-export class SideBarTreeProvider implements vscode.TreeDataProvider<NodeEntry>{
-
-    /*private configSetting: SettingNode;
-    private buildSetting: SettingNode;
-    private testSetting: SettingNode;
-    private project: SettingNode;*/
+export class SideBarTreeProvider implements TreeDataProvider<Node>{
 
-    private _onDidChangeTreeData: vscode.EventEmitter<any> = new vscode.EventEmitter<any>();
-	readonly onDidChangeTreeData: vscode.Event<any> = this._onDidChangeTreeData.event;
+    private _onDidChangeTreeData: EventEmitter<any> = new EventEmitter<any>();
+    readonly onDidChangeTreeData: Event<any> = this._onDidChangeTreeData.event;
 
-    constructor(){
+    constructor() {
     }
-    setProject (_project: CMakeProject): void {
+
+    updateActiveProject(cmakeProject?: CMakeProject): void {
         // Use project to create the tree
+        if (cmakeProject) {
+            Node.updateActiveProject(cmakeProject);
+            this.refresh();
+        }
     }
     public refresh(): any {
-		this._onDidChangeTreeData.fire(undefined);
-	}
-    getTreeItem(element: subNodeEntry): vscode.TreeItem {
-        if (element.nodeType === NodeType.Configure || element.nodeType === NodeType.Build || element.nodeType === NodeType.Test) {
-            return {
-                label: element.nodeType,
-                tooltip: "",
-                collapsibleState: vscode.TreeItemCollapsibleState.Expanded
-            };
+        this._onDidChangeTreeData.fire(undefined);
+    }
+    getTreeItem(element: Node): TreeItem {
+        return element.getTreeItem();
+    }
+    getChildren(element?: Node | undefined): Node[] {
+        // When initializing the tree
+        if (!element) {
+            return [new ConfigNode(), new BuildNode()];
         } else {
-            if (element.kit) {
-                const lable = element.kit;
-                const item = new vscode.TreeItem(lable);
-                //item.command = ;
-                const icon = 'binary-icon.svg';
-                item.iconPath = {
+            return element.getChildren();
+        }
+    }
+
+}
+
+enum NodeType {
+    Configure = "Configure",
+    Build = "Build",
+    Test = "Test",
+    Preset = "Preset",
+    Kit = "Kit",
+    Variant = "Variant",
+    Target = "Target"
+}
+
+class Node extends TreeItem{
+
+    static cmakeProject?: CMakeProject;
+    static updateActiveProject(cmakeProject?: CMakeProject): void {
+        // Use project to create the tree
+        if (cmakeProject) {
+            Node.cmakeProject = cmakeProject;
+        }
+    }
+
+    getTreeItem(): TreeItem {
+        return this;
+    }
+
+    getChildren(): Node[] {
+        return [];
+    }
+
+    async initialize(nodeType: NodeType): Promise<void> {
+        let icon: string;
+        if (!Node.cmakeProject) {
+            return;
+        }
+        switch (nodeType) {
+            case NodeType.Configure:
```

Comment:
For each of the cases in this `switch` the code should go in the constructor of the matching class.